Concatenation of tuples

Concatenation of fixed-length tuples should be supported by the type checker. 

Arbitrary tuples could be concatenated. `Tuple[int, str] + Tuple[bool]` results in `Tuple[int, str, bool]`.

Concatenating two homogeneous tuples (`Tuple[T, ...]`) already works.

Hi @JukkaL, any update on this feature ? Currently, concatenating two fixed-lengths tuples results in a `Unsupported operand types for +` error and a messed up type as demonstrated by the example below:

```
a = (1,) # Tuple[int]
b = ('bob',) # Tuple[str]
c = a + b # Tuple[int, str]
reveal_type(a)
reveal_type(b)
reveal_type(c)
```

```
$ mypy a.py 
a.py:5: error: Unsupported operand types for + ("Tuple[int]" and "Tuple[str]")
a.py:13: error: Revealed type is 'Tuple[builtins.int]'
a.py:14: error: Revealed type is 'Tuple[builtins.str]'
a.py:15: error: Revealed type is 'builtins.tuple[builtins.int*]'
```

--- a/mypy/checkexpr.py
+++ b/mypy/checkexpr.py
@@ -1886,6 +1886,13 @@ class ExpressionChecker(ExpressionVisitor[Type]):
                 column=typ.column,
             ))
 
+    def concat_tuples(self, left: TupleType, right: TupleType) -> TupleType:
+        """Concatenate two fixed length tuples."""
+        result = TupleType(items=left.items + right.items,
+                           fallback=self.named_type('builtins.tuple'))
+        result.partial_fallback = tuple_fallback(result)
+        return result
+
     def visit_int_expr(self, e: IntExpr) -> Type:
         """Type check an integer literal (trivial)."""
         return self.infer_literal_expr_type(e.value, 'builtins.int')
@@ -1944,6 +1951,12 @@ class ExpressionChecker(ExpressionVisitor[Type]):
                 if isinstance(e.left, (StrExpr, BytesExpr, UnicodeExpr)):
                     return self.strfrm_checker.check_str_interpolation(e.left, e.right)
         left_type = self.accept(e.left)
+
+        proper_left_type = get_proper_type(left_type)
+        if isinstance(proper_left_type, TupleType) and e.op == '+':
+            proper_right_type = get_proper_type(self.accept(e.right))
+            if isinstance(proper_right_type, TupleType):
+                return self.concat_tuples(proper_left_type, proper_right_type)
 
         if e.op in nodes.op_methods:
             method = self.get_operator_method(e.op)

--- a/mypy/checkexpr.py
+++ b/mypy/checkexpr.py
@@ -1952,9 +1952,13 @@ class ExpressionChecker(ExpressionVisitor[Type]):
 
         proper_left_type = get_proper_type(left_type)
         if isinstance(proper_left_type, TupleType) and e.op == '+':
-            proper_right_type = get_proper_type(self.accept(e.right))
-            if isinstance(proper_right_type, TupleType):
-                return self.concat_tuples(proper_left_type, proper_right_type)
+            left_add_method = proper_left_type.partial_fallback.type.get('__add__')
+            if left_add_method and left_add_method.fullname == 'builtins.tuple.__add__':
+                proper_right_type = get_proper_type(self.accept(e.right))
+                if isinstance(proper_right_type, TupleType):
+                    right_radd_method = proper_right_type.partial_fallback.type.get('__radd__')
+                    if right_radd_method is None:
+                        return self.concat_tuples(proper_left_type, proper_right_type)
 
         if e.op in nodes.op_methods:
             method = self.get_operator_method(e.op)

--- a/mypy/checkexpr.py
+++ b/mypy/checkexpr.py
@@ -1888,10 +1888,8 @@ class ExpressionChecker(ExpressionVisitor[Type]):
 
     def concat_tuples(self, left: TupleType, right: TupleType) -> TupleType:
         """Concatenate two fixed length tuples."""
-        result = TupleType(items=left.items + right.items,
-                           fallback=self.named_type('builtins.tuple'))
-        result.partial_fallback = tuple_fallback(result)
-        return result
+        return TupleType(items=left.items + right.items,
+                         fallback=self.named_type('builtins.tuple'))
 
     def visit_int_expr(self, e: IntExpr) -> Type:
         """Type check an integer literal (trivial)."""

--- a/test-data/unit/check-tuples.test
+++ b/test-data/unit/check-tuples.test
@@ -738,7 +738,7 @@ class C: pass
 
 a = None # type: A
 
-(a, a) + a  # E: Unsupported left operand type for + ("Tuple[A, A]")
+(a, a) + a  # E: Unsupported operand types for + ("Tuple[A, A]" and "A")
 a + (a, a)  # E: Unsupported operand types for + ("A" and "Tuple[A, A]")
 f((a, a))   # E: Argument 1 to "f" has incompatible type "Tuple[A, A]"; expected "A"
 (a, a).foo  # E: "Tuple[A, A]" has no attribute "foo"

--- a/test-data/unit/check-tuples.test
+++ b/test-data/unit/check-tuples.test
@@ -1239,3 +1239,5 @@ a = (1, "foo", 3)
 b = ("bar", 7)
 
 reveal_type(a + b)  # N: Revealed type is 'Tuple[builtins.int, builtins.str, builtins.int, builtins.str, builtins.int]'
+
+[builtins fixtures/tuple.pyi]

--- a/test-data/unit/check-tuples.test
+++ b/test-data/unit/check-tuples.test
@@ -1233,3 +1233,9 @@ reveal_type(tup[2])  # N: Revealed type is 'Union[Any, builtins.int*]' \
 reveal_type(tup[:])  # N: Revealed type is 'Union[Tuple[builtins.int, builtins.str], builtins.list[builtins.int*]]'
 
 [builtins fixtures/tuple.pyi]
+
+[case testFixedLengthTupleConcatenation]
+a = (1, "foo", 3)
+b = ("bar", 7)
+
+reveal_type(a + b)  # N: Revealed type is 'Tuple[builtins.int, builtins.str, builtins.int, builtins.str, builtins.int]'

--- a/test-data/unit/fixtures/tuple.pyi
+++ b/test-data/unit/fixtures/tuple.pyi
@@ -1,6 +1,6 @@
 # Builtins stub used in tuple-related test cases.
 
-from typing import Iterable, Iterator, TypeVar, Generic, Sequence, Any, overload
+from typing import Iterable, Iterator, TypeVar, Generic, Sequence, Any, overload, Tuple
 
 Tco = TypeVar('Tco', covariant=True)
 
@@ -15,6 +15,7 @@ class tuple(Sequence[Tco], Generic[Tco]):
     def __contains__(self, item: object) -> bool: pass
     def __getitem__(self, x: int) -> Tco: pass
     def __rmul__(self, n: int) -> tuple: pass
+    def __add__(self, x: Tuple[Tco, ...]) -> Tuple[Tco, ...]: pass
     def count(self, obj: Any) -> int: pass
 class function: pass
 class ellipsis: pass

